x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/formula/formula_public_api.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { IndexPattern, PersistedIndexPatternLayer } from '../../../types';
+import type { DataView } from '../../../../../../../../src/plugins/data_views/public';
+
+import { insertOrReplaceFormulaColumn } from '.';
+import { operationDefinitionMap } from '../../operations';
+import { convertDataViewIntoLensIndexPattern } from '../../../loader';
+
+/** @public **/
+export interface FormulaPublicApi {
+  /**
+   * Method which Lens consumer can import and given a formula string,
+   * return a parsed result as list of columns to use as Embeddable attributes.
+   *
+   * @param id - Formula column id
+   * @param column.formula - String representation of a formula
+   * @param [column.label] - Custom formula label
+   * @param layer - The layer to which the formula columns will be added
+   * @param dataView - The dataView instance
+   * @param [params.operations] - Use this parameter only if you need to filter available operations in the formula.
+   *
+   * See `x-pack/examples/embedded_lens_example` for exemplary usage.
+   */
+  insertOrReplaceFormulaColumn: (
+    id: string,
+    column: {
+      formula: string;
+      label?: string;
+    },
+    layer: PersistedIndexPatternLayer,
+    dataView: DataView,
+    params?: { operations?: string[] }
+  ) => PersistedIndexPatternLayer | undefined;
+}
+
+/** @public **/
+export const createFormulaPublicApi = (): FormulaPublicApi => {
+  const cache: WeakMap<DataView, IndexPattern> = new WeakMap();
+
+  const getCachedLensIndexPattern = (dataView: DataView): IndexPattern => {
+    const cachedIndexPattern = cache.get(dataView);
+    if (cachedIndexPattern) {
+      return cachedIndexPattern;
+    }
+    const indexPattern = convertDataViewIntoLensIndexPattern(dataView);
+    cache.set(dataView, indexPattern);
+    return indexPattern;
+  };
+
+  return {
+    insertOrReplaceFormulaColumn: (id, { formula, label }, layer, dataView, params) => {
+      const indexPattern = getCachedLensIndexPattern(dataView);
+
+      return insertOrReplaceFormulaColumn(
+        id,
+        {
+          label: label ?? formula,
+          customLabel: Boolean(label),
+          operationType: 'formula',
+          dataType: 'number',
+          references: [],
+          isBucketed: false,
+          params: {
+            formula,
+          },
+        },
+        { ...layer, indexPatternId: indexPattern.id },
+        {
+          indexPattern,
+          operations:
+            params?.operations?.reduce(
+              (operationsMap, item) => ({
+                ...operationsMap,
+                [item]: operationDefinitionMap[item],
+              }),
+              {}
+            ) ?? undefined,
```

Comment:
nit: I think this is not required. The only scenario where the right part of `??` is executed is actually to have the left part already `undefined`.